infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvVarsConverterTest.java

```diff
@@ -0,0 +1,108 @@
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalMachineConfig;
+import org.eclipse.che.workspace.infrastructure.kubernetes.Names;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+public class EnvVarsConverterTest {
+
+  private static final String PRE_EXISTING_VAR = "VAR_THAT_EXISTS";
+  private static final String PRE_EXISTING_VAR_VALUE = "jmenuju se VAR";
+  private static final String PRE_EXISTING_VAR_NEW_VALUE = "my name is VAR";
+  private static final String A_VAR = "A";
+  private static final String A_VAL = "$(C)";
+  private static final String B_VAR = "B";
+  private static final String B_VAL = "b";
+  private static final String C_VAR = "C";
+  private static final String C_VAL = "c";
+
+  private KubernetesEnvironment environment;
+  private RuntimeIdentity identity;
+  private Container testContainer;
+  private InternalMachineConfig machine;
+
+  @BeforeMethod
+  public void setUp() {
+    testContainer = new Container();
+
+    PodSpec podSpec = new PodSpec();
+    podSpec.setContainers(singletonList(testContainer));
+
+    ObjectMeta podMeta = new ObjectMeta();
+    podMeta.setName("pod");
+
+    Pod pod = new Pod();
+    pod.setSpec(podSpec);
+    pod.setMetadata(podMeta);
+
+    Map<String, Pod> pods = new HashMap<>();
+    pods.put("pod", pod);
+
+    environment = KubernetesEnvironment.builder().setPods(pods).build();
+
+    machine = new InternalMachineConfig();
+
+    environment.setMachines(
+        Collections.singletonMap(Names.machineName(podMeta, testContainer), machine));
+
+    identity = new RuntimeIdentityImpl("wsId", "blah", "bleh");
+  }
+
+  @Test
+  public void shouldProvisionEnvironmentVariablesSorted() throws InfrastructureException {
+    // given
+    List<EnvVar> preExistingEnvironment = new ArrayList<>();
+    preExistingEnvironment.add(new EnvVar(PRE_EXISTING_VAR, PRE_EXISTING_VAR_VALUE, null));
+    testContainer.setEnv(preExistingEnvironment);
+
+    machine.getEnv().put(PRE_EXISTING_VAR, PRE_EXISTING_VAR_NEW_VALUE);
+    machine.getEnv().put(A_VAR, A_VAL);
+    machine.getEnv().put(B_VAR, B_VAL);
+    machine.getEnv().put(C_VAR, C_VAL);
+
+    // when
+    EnvVarsConverter converter = new EnvVarsConverter();
+    converter.provision(environment, identity);
+
+    // then
+    EnvVar expectedA = new EnvVar(A_VAR, A_VAL, null);
+    EnvVar expectedB = new EnvVar(B_VAR, B_VAL, null);
+    EnvVar expectedC = new EnvVar(C_VAR, C_VAL, null);
+    EnvVar expectedPreExisting = new EnvVar(PRE_EXISTING_VAR, PRE_EXISTING_VAR_NEW_VALUE, null);
+
+    assertEquals(4, testContainer.getEnv().size());
+    assertEquals(expectedB, testContainer.getEnv().get(0));
+    assertEquals(expectedC, testContainer.getEnv().get(1));
```

Comment:
There is `assertEqualsNoOrder` for such cases. Unfortunately it check only arrays, but you can convert lists to arrays with `list.toArray()`. This would not change much, but the test would be a little bit more concise.

Reply:
`assertEqualsNoOrder` seems to not be available in TestNG assert. I made a change to the test code anyway so that the purpose is hopefully slightly more clear.

Reply:
Now it is cleaner, agree. But arrays assertion should be in test ng. See http://static.javadoc.io/org.testng/testng/6.0.1/org/testng/Assert.html#assertEqualsNoOrder(java.lang.Object[],%20java.lang.Object[])
Maybe it will be useful next time

Reply:
Huh, I'm blind :wink: Thanks for pointing it me to the method, I didn't know about it before. In any case, it would not be acceptable here anyway, because the test is specifically for the order of the elements, checking that the env vars are topologically sorted.

Reply:
Oh, yeah, you are right. I missed that we need to check the order